Fix makefile and test case in sslinfo

gfphoenix78 opened the pull request:
1. Make sslinfo compiled only if `--with-openssl` is provided
2. Fix sql scripts and answer files in sslinfo

- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

guofengrichard commented on GNUmakefile.in:
```diff
@@ -139,22 +143,17 @@ $(call recurse,check-world,src/test src/pl src/interfaces/ecpg contrib src/bin g
 # probably pull anything order-dependent out of recurse() and back into the
 # recipe body).
 # GPDB_94_MERGE_FIXME: We should stop listing each contrib module here separately.
```
This comment is out of date. Please modify/remove it.

gfphoenix78 replied:
Removed GPDB_94_MERGE_FIXME

paul-guo- commented:
I feel we should follow upstream behaviors,

i.e. do it in contrib/Makefile
```
  ifeq ($(with_openssl),yes)
  SUBDIRS += sslinfo
  else
  ALWAYS_SUBDIRS += sslinfo
  endif
```

Not sure why we added those contrib/ targets in GNUmakefile.in? Is it possible to modify the Makefiles to follow upstream? and then add the test case in the 2nd patch?

gfphoenix78 commented:
@paul-guo- 

> I feel we should follow upstream behaviors,
> 
> i.e. do it in contrib/Makefile
> 
> ```
>   ifeq ($(with_openssl),yes)
>   SUBDIRS += sslinfo
>   else
>   ALWAYS_SUBDIRS += sslinfo
>   endif
> ```
> 
> Not sure why we added those contrib/ targets in GNUmakefile.in? Is it possible to modify the Makefiles to follow upstream? and then add the test case in the 2nd patch?

I don't know what it exactly means in contrib/Makefile. When I run `configure` and `make`,  sslinfo  is not compiled at all, if I removed the lines about sslinfo in GNUMakefile.in. Note that the upstream does have the above code in contrib/Makefile

d commented on GNUmakefile.in:
```diff
@@ -59,6 +62,9 @@ install:
 	$(MAKE) -C contrib/pg_xlogdump $@
 	$(MAKE) -C contrib/hstore $@
 	$(MAKE) -C contrib/pgcrypto $@
+ifeq ($(with_openssl), yes)
```
No please no.

In fact, this **entire block of** mentions of `contrib` is an antipattern here. It seems to suggest we `make` (and `make install` to do something that's meant to be the job of `make world` and `make install-world`

danielgustafsson replied:
Yes and no. We want to provide a set of modules built and installed by default on `make install`, that is not something `contrib/Makefile` is intend to do. `contrib` is an opt-in where you either get a single module with `make -C  contrib/...` or everything with `make -C contrib`. I'm not saying we shouldn't change this, but changing what `make -C contrib` means wouldn't be an improvement when what we want is a cherry-picked default set.

d replied:
@danielgustafsson I don't recall any decision explicitly around *which* module goes into the curated set. I don't have a problem with having a curated set, to be clear. But what I _do_ have a problem with, is that our development process has become:

1. We run `make install` in CI
1. We then run `make installcheck-world`
1. People have come to believe -- reasonably -- that in order to get test coverage in CI, they _have to_ include their code in `make install`. It wouldn't even occur to most people that this inadvertently adds something to the "curated" set.

There is a real disparity between step one and step two, IMO. If we've standardized on running `installcheck-world`, wouldn't it be reasonable that we use `install-world`?

danielgustafsson replied:
Right, regarding this boiled down complaint we are in total agreement.

The curated set has more to do with product management than any technical reasoning. It is (was) intended as the default set of extra modules that goes into the final packaging, and it was hooked into the  top-level `GNUMakefile` some time in the past for reasons lost to history (simplicity I would assume). What we should do is create some way of building/installing the default curated list; make sure that whatever the curated list is it is *always* a subset of `make world`; and use `make install-world` + `make installcheck-world` in the CI.

Do you feel like proposing something along these lines?

d commented on contrib/sslinfo/config.bash:
```diff
@@ -0,0 +1,63 @@
+echo "ssl_ca_file='root.crt'" >> $MASTER_DATA_DIRECTORY/postgresql.conf
+echo "#END SSLINFO CONF : END ANCHOR##" >> $MASTER_DATA_DIRECTORY/postgresql.conf
+
+echo "#BEGIN SSLINFO CONF : BEGIN ANCHOR##" >> $standby_data/postgresql.conf
```
The repetition of `echo` (which itself is less portable) obscures what's being added to the configuration. There's also ~7 or 8 lines that's duplicated. Have you considered using a variable and here-doc?

```suggestion
local SSLINFO_CONF
SSLINFO_CONF="#BEGIN SSLINFO CONF: BEGIN ANCHOR ##
ssl=on
ssl_ciphers='HIGH:MEDIUM:+3DES:!aNULL'
ssl_cert_file='server.crt'
ssl_key_file='server.key'
ssl_ca_file='root.crt'
#END SSLINFO CONF : END ANCHOR##
"

cat >> $MASTER_DATA_DIRECTORY/postgresql.conf <<<$SSLINFO_CONF
cat >> $standby_data/postgresql.conf <<<$SSLINFO_CONF
```

gfphoenix78 replied:
Hi, Jesse, why is `echo` less portable? Is there any shell that doesn't support echo well? Compare to `echo`, I think multiple-line string is less portable, since `echo` can be seen everywhere, and multiple-line string is much less seen. On the other hand, the script file is explicitly executed by `bash`. It only causes problems if the system has no bash or has a fake bash. 
For a more general question, should `sh` be considered the default shell interpreter? `sh` may point to different shell in different system. But centos/ubuntu/suse12/ all have bash as we expected.

d commented on contrib/sslinfo/expected/sslinfo.out:
```diff
@@ -0,0 +1,130 @@
+(1 row)
+
+SELECT ssl_cipher() IS NOT NULL AS cipher_ok;
+ cipher_ok 
```
Seriously, how much is testing for `NOT NULL` better than not testing it?

gfphoenix78 replied:
`NOT NULL` means I **CAN** get the cipher info from the configuration, and it is **NOT** empty. It doesn't care what it is.

d commented on contrib/sslinfo/expected/sslinfo.out:
```diff
@@ -0,0 +1,130 @@
+ t
+(1 row)
+
+SELECT ssl_version() like 'TLSv%' AS version;
```
-1

No, please no.

This code does not seem to be testing much of `sslinfo`, instead it is testing the behavior of `SSL_get_version`, which is implemented in OpenSSL (or whatever provides an OpenSSL-like interface, e.g. GnuTLS libSSL adapter). There is no code in the `sslinfo` module that controls the output here, even the prefix of `TLS`.

gfphoenix78 replied:
Make sense.

gfphoenix78 commented:
Rebase branch and recommit due to conflict in GNUMakefile.in
